src/main/java/com/exactpro/th2/check1/Checkpoint.java

```diff
@@ -53,21 +57,34 @@ public long getSequence(SessionKey sessionKey) {
         return sessionKeyToSequence.get(sessionKey);
     }
 
-    public com.exactpro.th2.common.grpc.Checkpoint convert() {
-        Map<String, DirectionCheckpoint.Builder> intermediateMap = new HashMap<>();
-
-        for (Map.Entry<SessionKey, Long> entry : sessionKeyToSequence.entrySet()) {
-            SessionKey sessionKey = entry.getKey();
-            intermediateMap.computeIfAbsent(sessionKey.getSessionAlias(), alias -> DirectionCheckpoint.newBuilder())
-                    .putDirectionToSequence(sessionKey.getDirection().getNumber(), entry.getValue());
-        }
-
-        var checkpointBuilder = com.exactpro.th2.common.grpc.Checkpoint.newBuilder()
+    public com.exactpro.th2.common.grpc.Checkpoint toProto() {
+        var bookToAliasToDirectionToSequence = sessionKeyToSequence.entrySet().stream()
+                .collect(groupingBy(
+                                bookEntry -> bookEntry.getKey().getBookName(),
+                                groupingBy(
+                                        aliasEntry -> aliasEntry.getKey().getSessionAlias(),
+                                        toMap(
+                                                directionEntry -> directionEntry.getKey().getDirection().getNumber(),
+                                                Map.Entry::getValue
+                                        )
+                                )
+                        )
+                );
+
+        var checkpointBuilder = com.exactpro.th2.common.grpc.Checkpoint
+                .newBuilder()
                 .setId(id);
-        for (Map.Entry<String, DirectionCheckpoint.Builder> entry : intermediateMap.entrySet()) {
-            checkpointBuilder.putSessionAliasToDirectionCheckpoint(entry.getKey(),
-                    entry.getValue().build());
-        }
+        bookToAliasToDirectionToSequence.forEach((book, aliasToDirectionToSequence) -> {
```

Comment:
Maybe continue the pipeline above?

Reply:
Maybe you are right and old pipeline was more clear then `groupingBy(groupingBy(...))`.
I returned old pipeline and added book, please see.
Also I removed deprecated usages of `putDirectionToSequence()` and `getDirectionToSequenceMap()` and I use `CheckpointData` instead of `Long`.

Reply:
As discussed on meeting I'll get `bookName` and `scope` from `parentID` in `Event.toProto()`.
Also I'll add another method to `common.Event` like `toProto()` which will accept `bookName` and `scope`.